File: cpp/tests/test_diffusion.cpp
#include <biotransport/solvers/diffusion.hpp>
#include <biotransport/solvers/reaction_diffusion.hpp>
#include <biotransport/core/utils.hpp>
#include <vector>
#include <cassert>
#include <iostream>
#include <cmath>

// Test the diffusion solver with a known analytical solution
void testDiffusion1D() {
    std::cout << "Testing 1D diffusion solver..." << std::endl;
    
    // Create a 1D mesh
    biotransport::StructuredMesh mesh(100, 0.0, 1.0);
    
    // Create a diffusion solver
    double D = 0.01;  // diffusion coefficient
    biotransport::DiffusionSolver solver(mesh, D);
    
    // Set up initial condition (step function)
    std::vector<double> initial(mesh.numNodes(), 0.0);
    for (int i = 0; i <= mesh.nx(); ++i) {
        double x = mesh.x(i);
        initial[i] = (x >= 0.4 && x <= 0.6) ? 1.0 : 0.0;
    }
    solver.setInitialCondition(initial);
    
    // Set boundary conditions
    solver.setDirichletBoundary(0, 0.0);
    solver.setDirichletBoundary(1, 0.0);
    
    // Solve
    double dt = 0.0001;
    int num_steps = 1000;
    solver.solve(dt, num_steps);
    
    // Get the solution
    const auto& solution = solver.solution();
    
    // Check that the solution is reasonable
    // - Mass conservation (approximately)
    double initial_mass = 0.0;
    double final_mass = 0.0;
    for (int i = 0; i <= mesh.nx(); ++i) {
        initial_mass += initial[i] * mesh.dx();
        final_mass += solution[i] * mesh.dx();
    }
    assert(std::abs(final_mass - initial_mass) < 0.1);  // Some loss expected due to boundaries
    
    // - Symmetry (since initial condition and BCs are symmetric)
    for (int i = 0; i <= mesh.nx() / 2; ++i) {
        assert(std::abs(solution[i] - solution[mesh.nx() - i]) < 1e-6);
    }
    
    // - Peak value decreased from initial
    double max_val = 0.0;
    for (int i = 0; i <= mesh.nx(); ++i) {
        max_val = std::max(max_val, solution[i]);
    }
    assert(max_val < 1.0);
    
    std::cout << "1D diffusion tests passed!" << std::endl;
}

// Test the reaction-diffusion solver with a simple decay reaction
void testReactionDiffusion1D() {
    std::cout << "Testing 1D reaction-diffusion solver..." << std::endl;
    
    // Create a 1D mesh
    biotransport::StructuredMesh mesh(100, 0.0, 1.0);
    
    // Define reaction term (decay)
    double k = 0.01;  // decay rate
    auto reaction = [k](double u, double x, double y, double t) {
        return -k * u;
    };
    
    // Create a reaction-diffusion solver
    double D = 0.01;  // diffusion coefficient
    biotransport::ReactionDiffusionSolver solver(mesh, D, reaction);
    
    // Set up initial condition (step function)
    std::vector<double> initial(mesh.numNodes(), 0.0);
    for (int i = 0; i <= mesh.nx(); ++i) {
        double x = mesh.x(i);
        initial[i] = (x >= 0.4 && x <= 0.6) ? 1.0 : 0.0;
    }
    solver.setInitialCondition(initial);
    
    // Set boundary conditions
    solver.setDirichletBoundary(0, 0.0);
    solver.setDirichletBoundary(1, 0.0);
    
    // Solve
    double dt = 0.0001;
    int num_steps = 1000;
    solver.solve(dt, num_steps);
    
    // Get the solution
    const auto& solution = solver.solution();
    
    // Check that the solution is reasonable
    // - Total mass should be less than initial due to decay
    double initial_mass = 0.0;
    double final_mass = 0.0;
    for (int i = 0; i <= mesh.nx(); ++i) {
        initial_mass += initial[i] * mesh.dx();
        final_mass += solution[i] * mesh.dx();
    }
    assert(final_mass < initial_mass);  // Mass should decrease due to decay
    
    // - Symmetry (since initial condition and BCs are symmetric)
    for (int i = 0; i <= mesh.nx() / 2; ++i) {
        assert(std::abs(solution[i] - solution[mesh.nx() - i]) < 1e-6);
    }
    
    // - Peak value decreased from initial
    double max_val = 0.0;
    for (int i = 0; i <= mesh.nx(); ++i) {
        max_val = std::max(max_val, solution[i]);
    }
    assert(max_val < 1.0);
    
    std::cout << "1D reaction-diffusion tests passed!" << std::endl;
}

int main() {
    // Run tests
    testDiffusion1D();
    testReactionDiffusion1D();
    
    std::cout << "All diffusion tests passed!" << std::endl;
    return 0;
}

File: python/tests/test_diffusion.py
import unittest
import numpy as np
from biotransport import StructuredMesh, DiffusionSolver, ReactionDiffusionSolver


class TestDiffusion(unittest.TestCase):
    def test_1d_mesh(self):
        """Test 1D mesh creation and properties."""
        mesh = StructuredMesh(10, 0.0, 1.0)

        self.assertEqual(mesh.nx(), 10)
        self.assertTrue(mesh.is_1d())
        self.assertEqual(mesh.num_nodes(), 11)
        self.assertEqual(mesh.num_cells(), 10)
        self.assertAlmostEqual(mesh.dx(), 0.1)

        # Test node coordinates
        self.assertAlmostEqual(mesh.x(0), 0.0)
        self.assertAlmostEqual(mesh.x(5), 0.5)
        self.assertAlmostEqual(mesh.x(10), 1.0)

    def test_2d_mesh(self):
        """Test 2D mesh creation and properties."""
        mesh = StructuredMesh(5, 5, 0.0, 1.0, 0.0, 1.0)

        self.assertEqual(mesh.nx(), 5)
        self.assertEqual(mesh.ny(), 5)
        self.assertFalse(mesh.is_1d())
        self.assertEqual(mesh.num_nodes(), 36)
        self.assertEqual(mesh.num_cells(), 25)
        self.assertAlmostEqual(mesh.dx(), 0.2)
        self.assertAlmostEqual(mesh.dy(), 0.2)

        # Test node coordinates
        self.assertAlmostEqual(mesh.x(0), 0.0)
        self.assertAlmostEqual(mesh.x(5), 1.0)
        self.assertAlmostEqual(mesh.y(0, 0), 0.0)
        self.assertAlmostEqual(mesh.y(0, 5), 1.0)

    def test_diffusion_solver(self):
        """Test basic diffusion solver functionality."""
        mesh = StructuredMesh(50, 0.0, 1.0)
        D = 0.01
        solver = DiffusionSolver(mesh, D)

        # Create initial condition (Gaussian)
        x = np.array([mesh.x(i) for i in range(mesh.nx() + 1)])
        initial = np.exp(-100 * (x - 0.5) ** 2)
        solver.set_initial_condition(initial)

        # Set boundary conditions
        solver.set_dirichlet_boundary(0, 0.0)
        solver.set_dirichlet_boundary(1, 0.0)

        # Solve
        dt = 0.0001
        num_steps = 100
        solver.solve(dt, num_steps)

        # Get solution
        solution = solver.solution()

        # Basic checks
        self.assertEqual(len(solution), mesh.num_nodes())
        self.assertLess(max(solution), max(initial))  # Peak should decrease
        self.assertGreater(max(solution), 0.0)  # Solution shouldn't be all zero

    def test_reaction_diffusion_solver(self):
        """Test reaction-diffusion solver with a decay term."""
        mesh = StructuredMesh(50, 0.0, 1.0)
        D = 0.01

        # Define decay reaction
        k = 0.1  # decay rate

        def reaction(u, x, y, t):
            return -k * u

        solver = ReactionDiffusionSolver(mesh, D, reaction)

        # Create initial condition (Gaussian)
        x = np.array([mesh.x(i) for i in range(mesh.nx() + 1)])
        initial = np.exp(-100 * (x - 0.5) ** 2)
        solver.set_initial_condition(initial)

        # Set boundary conditions
        solver.set_dirichlet_boundary(0, 0.0)
        solver.set_dirichlet_boundary(1, 0.0)

        # Solve
        dt = 0.0001
        num_steps = 100
        solver.solve(dt, num_steps)

        # Get solution
        solution = solver.solution()

        # Basic checks
        self.assertEqual(len(solution), mesh.num_nodes())
        self.assertLess(max(solution), max(initial))  # Peak should decrease

        # Compare with diffusion-only solution
        diffusion_solver = DiffusionSolver(mesh, D)
        diffusion_solver.set_initial_condition(initial)
        diffusion_solver.set_dirichlet_boundary(0, 0.0)
        diffusion_solver.set_dirichlet_boundary(1, 0.0)
        diffusion_solver.solve(dt, num_steps)
        diffusion_solution = diffusion_solver.solution()

        # Reaction-diffusion solution should have lower values due to decay
        self.assertLess(sum(solution), sum(diffusion_solution))


if __name__ == "__main__":
    unittest.main()
